google/cloud/forseti/scanner/audit/kms_rules_engine.py

```diff
@@ -0,0 +1,471 @@
+
+    def __init__(self, rule_name, rule_index, rule):
+        """Initialize.
+
+        Args:
+            rule_name (str): Name of the loaded rule.
+            rule_index (int): The index of the rule from the rule definitions.
+            rule (dict): The rule definition from the file.
+        """
+        self.rule_name = rule_name
+        self.rule_index = rule_index
+        self.rule = rule
+
+    def exceeds_rotation_schedule(self, creation_time, scan_time):
+        """Check if the key has been rotated within the time speciifed in the
+         policy.
+
+        Args:
+            creation_time (datetime): The time at which the primary version of
+            the key was created.
+            scan_time (datetime): Snapshot timestamp.
+
+        Returns:
+            bool: Returns true if key was not rotated within the time specified
+            in the policy.
+        """
+        crypto_key = self.rule['key']
+        for key_data in crypto_key:
+            key_rotation_period = key_data.get('rotation_period')
+        LOGGER.debug('Formatting rotation time...')
+        creation_time = creation_time[:-5]
+        formatted_last_rotation_time = datetime.datetime.strptime(
+            creation_time, string_formats.TIMESTAMP_MICROS)
+        last_rotation_time = (scan_time - formatted_last_rotation_time).days
+
+        if last_rotation_time <= key_rotation_period:
+            return False
+
+        return True
+
+    def find_violations(self, key):
+        """Find crypto key violations based on the max_age.
+
+        Args:
+            key (Resource): The resource to check for violations.
+
+        Returns:
+            list: Returns a list of RuleViolation named tuples
+        """
+        violations = []
+        creation_time = key.primary_version.get('createTime')
+        scan_time = date_time.get_utc_now_datetime()
+
+        if self.rule['mode'] == BLACKLIST:
+            if self.exceeds_rotation_schedule(creation_time,
+                                              scan_time):
+                violation_reason = ('Key %s was not rotated since %s.' %
+                                    (key.name, creation_time))
+                violations.append(RuleViolation(
+                    resource_id=key.id,
+                    resource_type=key.type,
+                    resource_name=key.name,
+                    full_name=key.crypto_key_type,
+                    rule_index=self.rule_index,
+                    rule_name=self.rule_name,
+                    violation_type=VIOLATION_TYPE,
+                    primary_version=key.primary_version,
+                    next_rotation_time=key.next_rotation_time,
+                    rotation_period=key.rotation_period,
+                    violation_reason=violation_reason,
+                    key_creation_time=key.create_time,
+                    version_creation_time=key.primary_version.get('createTime'),
```

Comment:
You created the creation_time variable above, you should be either using that or not creating that variable at all.

creation_time = key.primary_version.get('createTime')

Reply:
Good observation, fixed it.